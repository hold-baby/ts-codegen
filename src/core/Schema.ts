import { isArray, toCapitalCase } from "src/core/utils";
import { forEach, isEmpty, map, reduce } from "lodash";
import { IReference, ISchema } from "src/v3/OpenAPI";
import { CustomSchema, CustomType, Type } from "src/core/Type";

export class Schema {
  convert(schema: CustomSchema, name?: string): CustomType {
    const oneOf = (schema as ISchema).oneOf || (schema as ISchema).anyOf;
    if (oneOf) {
      return Type.oneOf(map(oneOf, (v) => this.convert(v)));
    }

    const allOf = (schema as ISchema).allOf;
    if (allOf) {
      const { props, extend } = this.handleAllOf(allOf, name);
      return Type.object(props, extend);
    }

    if (schema.items) {
      return Type.array(this.handleItems(schema.items, name));
    }

    if (schema.$ref) {
      return Type.ref((schema as IReference).$ref);
    }

    if (schema.enum) {
      return Type.enum(schema.enum, name);
    }

    if (schema.type === "object") {
      return Type.object(this.handleObject(schema, name)); // TODO: handle when schema.properties not exists
    }

    if (schema.type === "string") {
      return Type.string();
    }

    if (schema.type === "boolean") {
      return Type.boolean();
    }

    if (schema.type === "integer" || schema.type === "number") {
      return Type.number();
    }

    if (schema.type === "file") {
      return Type.file();
    }

    return Type.null();
  }

  handleAllOf(schemas: Array<CustomSchema>, name?: string) {
    const extend: any[] = [];
    let props: any = {};

    forEach(schemas, (schema) => {
      if (schema.$ref) {
        extend.push(this.convert(schema, name));
      } else if (!isEmpty(schema)) {
        props = this.convert(schema, name);
      }
    });

    return {
      extend,
      props,
    };
  }

  handleItems(items: CustomSchema | IReference | CustomSchema[], name?: string): CustomType | CustomType[] {
    if (isArray(items)) {
      return map(items, (v) => this.handleItems(v, name)) as CustomType[];
    }
    return this.convert(items, name);
  }

  handleObject(schema: CustomSchema, name?: string): { [key: string]: CustomType } {
    return reduce(
      schema.properties!,
      (res, v, k) => {
        const isRequired = (v as CustomSchema)?.required || schema.required?.includes(k);
        return {
          ...res,
          [`${k}${isRequired ? "" : "?"}`]: this.convert(v, `${toCapitalCase(name)}${toCapitalCase(k)}`),
        };
      },
      {},
    );
  }
}
