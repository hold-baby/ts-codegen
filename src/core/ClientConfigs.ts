import { camelCase, chain, compact, map, pick } from "lodash";
import { getRefId } from "src/core/utils";
import { Register } from "src/core/Register";
import { CustomOperation, CustomParameter, CustomParameters, CustomPath, CustomReference } from "src/core/types";

export const getOperations = (path: CustomPath) =>
  pick(path, ["get", "post", "put", "delete", "patch", "head"]) as { [method: string]: CustomOperation };

export const pickParams = (params?: CustomParameters) => (type: "path" | "query" | "body" | "formData") => {
  const list = map(params, (param: CustomParameter | CustomReference) => {
    let data = param;

    if ((param as CustomReference).$ref) {
      const name = getRefId((param as CustomReference).$ref);
      data = Register.parameters[name];
    }

    if ((data as CustomParameter).in === type) {
      return data;
    }
  });

  return compact(list);
};

export const getRequestURL = (pathName: string, basePath?: string) => {
  const isPathParam = (str: string) => str.startsWith("{");
  const resolvedPathName = chain(pathName)
    .split("/")
    .map((p) => (isPathParam(p) ? `$${p}` : p))
    .join("/")
    .value();

  return `${basePath}${resolvedPathName === "/" && !!basePath ? "" : resolvedPathName}`;
};

export const getOperationId = (operationId?: string) => camelCase(operationId);
