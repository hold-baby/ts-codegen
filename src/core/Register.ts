// 利用闭包持有状态（私有变量）
import { CustomType, Ref } from "src/core/Type";
import { Parameter } from "swagger-schema-official";
import { IReference, IRequestBody, IResponse } from "src/v3/OpenAPI";
import { get, set } from "lodash";

interface IStore {
  decls: { [id: string]: CustomType };
  refs: { [id: string]: Ref };
  prefixes: { [id: string]: string };
  parameters: {
    [id: string]: Parameter;
  };
  responses: {
    [id: string]: Response | IResponse | IReference;
  };
  requestBodies: {
    [id: string]: IReference | IRequestBody;
  };
}

export const createRegister = () => {
  const store: IStore = {
    decls: {},
    refs: {},
    prefixes: {},
    parameters: {},
    responses: {},
    requestBodies: {},
  };

  return {
    setType: (id: string, type: CustomType) => {
      store.decls[id] = type;
    },

    setPrefix: (id: string, prefix: string) => {
      store.prefixes[id] = prefix;
    },

    setParameter: (id: string, parameter: Parameter) => {
      store.parameters[id] = parameter;
    },

    setResponses: (id: string, response: Response | IResponse | IReference) => {
      store.responses[id] = response;
    },

    setRequestBody: (id: string, requestBody: IReference | IRequestBody) => {
      store.requestBodies[id] = requestBody;
    },

    setRef: (id: string) => {
      if (store.refs[id]) {
        return store.refs[id];
      }

      const type = new Ref(id);
      store.refs[id] = type;

      return type;
    },
    renameAllRefs: (cb: (newName: string) => string) => {
      for (let name in store.refs) {
        store.refs[name].rename(cb(name));
      }
    },
    getRefs() {
      return store.refs;
    },
    getDecls() {
      return store.decls;
    },
    getPrefixes() {
      return store.prefixes;
    },
    getData(paths: string[]) {
      return get(store, paths);
    },
    setData(paths: string[], data: any) {
      return set(store, paths, data);
    },
  };
};
