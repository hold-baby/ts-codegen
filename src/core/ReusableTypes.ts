import { IOpenAPI } from "src/v3/OpenAPI";
import {
  addPrefixForInterface,
  addPrefixForType,
  handleEnums,
  isArray,
  isObject,
  SchemaHandler,
  toCapitalCase,
  toTypes,
} from "src";
import { compact, Dictionary, forEach, includes, isEmpty, map, mapValues, replace } from "lodash";
import { ENUM_SUFFIX } from "src/core/constants";
import { Spec } from "swagger-schema-official";
import qs from "querystring";

const getQueryObjFromStr = (str: string) => {
  const item = str.replace(/\[\]/gi, "");
  const input = item.split("?")[1];
  if (input) {
    return {
      ...qs.parse(input),
      others: str.replace(`?${input}`, ""),
    };
  }
  return {} as any;
};

interface IData {
  _kind: string;
  _name: string;
  _value: string | Dictionary<any>;
}

interface IAllData {
  [key: string]: IData;
}

const handleStr = (str: string, allData: IAllData) => {
  const { type, name, others } = getQueryObjFromStr(str);
  if (type === "ref") {
    return others ? `${allData[name]._name}${others}` : allData[name]._name;
  }
  return str;
};

export const resolve = (input: string | Dictionary<any>, allData: IAllData): any => {
  if (typeof input == "string") {
    return handleStr(input, allData);
  }
  if (!isEmpty(input?._oneOf)) {
    return map(input?._oneOf, (v) => resolve(v, allData)).join("|");
  }
  if (isArray(input)) {
    return map(input, (v) => resolve(v, allData));
  }
  if (isObject(input)) {
    return mapValues(input, (item) => resolve(item, allData));
  }
  return "";
};

export class ReusableTypes {
  public resolvedSchemas: Dictionary<any> = {};

  static of(spec: Spec | IOpenAPI) {
    return new ReusableTypes(spec);
  }

  constructor(private spec: Spec | IOpenAPI) {}

  gen = (withPrefix: boolean = true) => {
    const schemaHandler = SchemaHandler.of((k, v) => {
      if (includes(k, ENUM_SUFFIX)) {
        const enumName = replace(k, ENUM_SUFFIX, "");
        this.resolvedSchemas[enumName] = {
          _kind: "enum",
          _name: enumName,
          _value: v,
        };
        return;
      }

      if (typeof v === "string" && v !== "object") {
        const name = toCapitalCase(k);
        this.resolvedSchemas[name] = {
          _kind: "type",
          _name: withPrefix ? addPrefixForType(name) : name,
          _value: v,
        };
        return;
      }

      const name = toCapitalCase(k);
      this.resolvedSchemas[name] = {
        _kind: "interface",
        _name: withPrefix ? addPrefixForInterface(name) : name,
        _value: v,
      };
    });

    const schemas = this.spec.definitions || (this.spec as IOpenAPI).components?.schemas;

    forEach(schemas, (v, k) => {
      schemaHandler.resolve({
        ...v,
        _name: k,
        _propKey: k,
      });
    });

    // console.log(this.resolvedSchemas, "this.resolvedSchemas");
    const resolvedSchemas = mapValues(this.resolvedSchemas, (item) => {
      return {
        _kind: item._kind,
        _name: item._name,
        _values: resolve(item._value, this.resolvedSchemas),
      };
    });

    const arr = Object.keys(resolvedSchemas)
      .sort()
      .map((key) => {
        const { _name, _kind, _values } = resolvedSchemas[key];

        if (_values === "object") {
          return `export interface ${_name} {[key:string]:any}`;
        }

        if (_kind === "enum") {
          return handleEnums(_values, _name);
        }

        if (!isEmpty(_values?._extends)) {
          return `export interface ${_name} extends ${_values?._extends.join(",")} ${toTypes(_values?._others)} `;
        }

        if (!isEmpty(_values?._oneOf)) {
          return `export type ${_name} = ${map(_values?._oneOf, (item) => toTypes(item)).join("|")}`;
        }

        if (_kind === "type") {
          return `export type ${_name} = ${_values}`;
        }

        if (_kind === "interface") {
          const types = toTypes(resolvedSchemas[key]._values);
          return `export interface ${_name} ${types}`;
        }
      });

    return {
      output: compact(arr),
      resolvedSchemas,
    };
  };
}
