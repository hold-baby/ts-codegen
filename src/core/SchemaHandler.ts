import { Schema } from "swagger-schema-official";
import { generateEnumType, isArray, isNumber, toCapitalCase } from "src/core/utils";
import { forEach, indexOf, map, reduce, some, last, isEmpty } from "lodash";
import { ISchema } from "src/v3/OpenAPI";
import qs from "querystring";

type TDictionary<T> = { [key: string]: T };
type TCustomSchema = (Schema | ISchema) & { _propKey?: string; _name?: string };
type TWriteTo = (k: string, v: any) => void;

export class SchemaHandler {
  static of(writeTo: TWriteTo) {
    return new SchemaHandler(writeTo);
  }

  constructor(public writeTo: TWriteTo) {}

  resolve = (schema: TCustomSchema = {}) => {
    this.writeTo(schema._name!, this.toType(schema));
  };

  toType = (schema: TCustomSchema = {}): TDictionary<any> | string => {
    const oneOf = (schema as ISchema).oneOf;
    if (oneOf) {
      return this.toOneOfType(oneOf);
    }

    const anyOf = (schema as ISchema).anyOf;
    if (anyOf) {
      return this.toOneOfType({
        ...anyOf,
        _name: schema._name,
      });
    }

    const allOf = (schema as ISchema).allOf;
    if (allOf) {
      return this.toAllOfType(allOf, schema._name);
    }

    if (schema.$ref) {
      return this.toRefType(schema);
    }

    if (schema.items) {
      return this.toArrayType(schema);
    }

    if (schema.enum) {
      return this.toEnumType(schema);
    }

    if (schema.type === "object") {
      return this.toObjectType(schema);
    }

    if (schema?.properties) {
      return this.toObjectType(schema);
    }

    if (schema.type === "integer") {
      return "number";
    }

    if (schema.type === "file") {
      return "File";
    }

    return schema.type || "";
  };

  toRefType = (schema: Schema | ISchema): string => {
    const getTypeByRef = (str?: string) => {
      if (!str) {
        return;
      }
      const list = str.split("/");
      return list[list.length - 1];
    };

    return `?name=${toCapitalCase(getTypeByRef(schema.$ref))}&${qs.stringify({
      type: "ref",
    })}`;
  };

  toArrayType = (schema: TCustomSchema) => {
    if (isArray(schema.items)) {
      return map(schema.items, (item: Schema | ISchema) =>
        this.toType({
          ...item,
          _name: schema._name,
          _propKey: schema._propKey,
        }),
      );
    }

    const itemType = this.toType({
      ...(schema.items as TCustomSchema),
      _name: schema._name,
      _propKey: schema._propKey,
    });

    return schema.type === "array" ? `${itemType}[]` : itemType;
  };

  toEnumType = (schema: TCustomSchema) => {
    const enumType = generateEnumType(schema._name, schema._propKey);
    const hasNumber = some(schema.enum, (v) => isNumber(v));

    this.writeTo(enumType, schema.enum);

    if (hasNumber) {
      return enumType;
    }

    return `keyof typeof ${enumType}`;
  };

  toObjectType = (schema: TCustomSchema): TDictionary<any> | string => {
    const handleProperties = () =>
      reduce(
        schema.properties,
        (o, v, k) => {
          return {
            ...o,
            [`${k}${indexOf(schema.required, k) > -1 ? "" : "?"}`]: this.toType({
              ...v,
              _propKey: k,
              _name: schema._name,
            }),
          };
        },
        {} as any,
      );
    return schema.properties ? handleProperties() : schema.type;
  };

  toOneOfType = (schemas: TCustomSchema) => ({
    _oneOf: map(schemas, (schema) =>
      this.toType({
        ...schema,
        _name: schemas._name,
      }),
    ),
  });

  toAllOfType = (schemas: Array<TCustomSchema>, _name?: string) => {
    const _extends: any[] = [];
    let _others = {};

    const lastSchema = last(schemas);
    if (schemas.length == 1 || isEmpty(lastSchema)) {
      return this.toType({
        ...schemas[0],
        _name,
      });
    }

    forEach(schemas, (schema) => {
      if (schema.$ref) {
        _extends.push(
          this.toType({
            ...schema,
            _name,
          }),
        );
      } else if (schema.type === "object") {
        _others = this.toType({
          ...schema,
          _name,
        });
      }
    });

    return {
      _extends,
      _others,
    } as TDictionary<any>;
  };
}
