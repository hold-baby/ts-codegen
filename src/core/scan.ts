import { IReference } from "src/v3/OpenAPI";
import { CustomSchema, Ref, Register } from "src/core/Type";
import { keys } from "lodash";
import { getUseExtends, Schema } from "src/core/Schema";
import { prettifyCode, toCapitalCase } from "src/core/utils";

const getDeclarationType = (schema: CustomSchema) => {
  if (schema.type === "object" || schema.properties || (schema.allOf && getUseExtends(schema.allOf))) {
    return "interface";
  }
  return "type";
};

const addPrefix = (name: string) => `${Register.prefixes[name] === "interface" ? "I" : "T"}${name}`;

export const scan = (schemas: { [k: string]: CustomSchema | IReference }) => {
  const schemaHandler = new Schema();

  keys(schemas).forEach((k) => {
    const name = toCapitalCase(k);
    const type = schemaHandler.convert(schemas[k], name);
    Register.setType(name, type);
    Register.setPrefix(name, getDeclarationType(schemas[k]));
  });

  for (let name in Register.refs) {
    (Register.refs[name] as Ref).rename(addPrefix(name));
  }

  let output = "";
  keys(Register.decls)
    .sort()
    .forEach((k) => {
      output =
        output +
        `export ${Register.prefixes[k]} ${addPrefix(k)} ${
          Register.prefixes[k] === "interface" ? "" : "="
        } ${Register.decls[k].toType()}${Register.prefixes[k] === "type" ? ";" : ""}\n\n`;
    });

  return prettifyCode(output);
};
