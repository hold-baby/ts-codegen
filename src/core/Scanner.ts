import { IOpenAPI, IReference } from "src/v3/OpenAPI";
import { CustomSchema, CustomType, Enum, Ref } from "src/core/Type";
import { compact, get, isEmpty, keys, mapValues, sortBy } from "lodash";
import { Schema } from "src/core/Schema";
import { getUseExtends, prettifyCode, setDeprecated, toCapitalCase, toTypes } from "src/core/utils";
import { Spec } from "swagger-schema-official";
import { IClientConfig } from "src/core/types";
import { ClientConfigsV3, getClientConfigsV2 } from "src";
import { Register } from "src/core/Register";

export const getDeclarationType = (schema: CustomSchema) => {
  if (schema.type === "object" || schema.properties || (schema.allOf && getUseExtends(schema.allOf))) {
    return "interface";
  }
  return "type";
};

const addPrefix = (name: string) => `${Register.prefixes[name] === "interface" ? "I" : "T"}${name}`;

export const getOutput = (): string => {
  let output = "";
  keys(Register.decls)
    .sort()
    .forEach((k) => {
      const t = Register.decls[k];
      if (t instanceof Enum) {
        output = output + `export ${t.toType()}\n\n`;
        return;
      }

      output =
        output +
        `export ${Register.prefixes[k]} ${addPrefix(k)} ${
          Register.prefixes[k] === "interface" ? "" : "="
        } ${t.toType()}${Register.prefixes[k] === "type" ? ";" : ""}\n\n`;
    });
  return output;
};

type KType = { [key: string]: CustomType };

export class Scanner {
  constructor(private spec: Spec | IOpenAPI) {}

  public scan(): string {
    // TODO: handle v3 base path later
    const basePath = this.spec.basePath || "";
    this.toReusableTypes(this.spec.definitions || (this.spec as IOpenAPI)?.components?.schemas);
    let clientConfigs: IClientConfig[] = this.spec.swagger
      ? getClientConfigsV2(this.spec.paths, basePath)
      : new ClientConfigsV3(this.spec.paths, basePath).clientConfigs;

    for (let name in Register.refs) {
      (Register.refs[name] as Ref).rename(addPrefix(name));
    }

    return prettifyCode(`${this.toRequest(clientConfigs)} \n\n ${getOutput()}`);
  }

  private toReusableTypes(schemas: { [k: string]: CustomSchema | IReference }) {
    const schemaHandler = new Schema();

    return keys(schemas).map((k) => {
      const name = toCapitalCase(k);
      const type = schemaHandler.convert(schemas[k], name);
      Register.setType(name, type);
      Register.setPrefix(name, getDeclarationType(schemas[k]));
      return type;
    });
  }

  private toRequest(clientConfigs: IClientConfig[]): string {
    // for (let name in Register.refs) {
    //   if (!(Register.refs[name] as Ref).alias) {
    //     (Register.refs[name] as Ref).rename(addPrefix(name));
    //   }
    // }
    const clientConfig = sortBy(clientConfigs, (o) => o.operationId);

    // TODO: refactor code later
    function mapper(obj: KType | { [key: string]: KType }): any {
      return toTypes(
        mapValues(obj, (v: any) => {
          if (!v.toType) {
            return mapper(v);
          }
          return v.toType(false);
        }),
      );
    }

    return clientConfig
      .map((v: IClientConfig) => {
        const TReq = !isEmpty(v.TReq) ? mapper(v.TReq as any) : "";
        const requestParamList = compact([...v.pathParams, ...v.queryParams, v.contentType ? "requestBody" : ""]);
        const requestInputs = isEmpty(requestParamList) ? "" : toRequestParams(requestParamList);

        const getParams = () => {
          const params = toRequestParams(get(v, "queryParams"));
          return params ? `params: ${params},` : "";
        };

        const getHeaders = () => (v.contentType ? `headers: { "Content-Type": '${v.contentType}' },` : "");

        const types = compact([TReq, v.TResp?.toType(false)]).join(",");
        return `
${v.deprecated ? setDeprecated(v.operationId) : ""}
export const ${v.operationId} = createRequestAction${types ? "<" + types + ">" : ""}("${
          v.operationId
        }", (${requestInputs}) => ({ url: \`${v.url}\`,method: "${v.method}",${
          v.contentType ? `data: requestBody,` : ""
        }${getParams()}${getHeaders()} })
);
`;
      })
      .join("\n\n");
  }
}

const toRequestParams = (data: any[] = []) =>
  !isEmpty(data)
    ? `{
${data.join(",\n")}
}`
    : undefined;
