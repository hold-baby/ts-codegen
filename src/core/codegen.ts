import { PathsResolverV3 } from "src/v3/PathsResolverV3";
import { ReusableTypes } from "src/core/ReusableTypes";
import { prettifyCode, testJSON } from "src/core/utils";
import axios from "axios";
import { map } from "lodash";
import { ERROR_MESSAGES } from "src/core/constants";
import * as fs from "fs";
import * as path from "path";
import { IOpenAPI } from "src/v3/OpenAPI";
import { Spec } from "swagger-schema-official";
import { PathsResolverV2 } from "src";

export const codegen = (spec: IOpenAPI | Spec): string => {
  if (!spec) {
    return "";
  }

  if ((spec as IOpenAPI).openapi) {
    return [
      ...PathsResolverV3.of(spec.paths, spec.basePath)
        .scan()
        .toRequest(),
      ...ReusableTypes.of(spec).gen(),
    ].join("\n\n");
  }

  return [
    ...PathsResolverV2.of(spec.paths, spec.basePath)
      .scan()
      .toRequest(),
    ...ReusableTypes.of(spec).gen(),
  ].join("\n\n");
};

const getFilename = (basePath?: string) =>
  basePath
    ? `./${basePath
        .split("/")
        .join(".")
        .slice(1)}`
    : "./api.client";

const write = (output: string, filename: string, str: string) => {
  if (!fs.existsSync(output)) {
    fs.mkdirSync(output);
  }

  fs.writeFileSync(path.resolve(output, `./${filename}.ts`), prettifyCode(str), "utf-8");
};

// const codegenFromRemote = () => {};

export const fetchSwaggerJSON = (clients: string[] = [], timeout: number = 10 * 1000) => {
  const instance = axios.create({
    timeout,
  });

  return Promise.all(
    map(clients, (client) => {
      instance
        .get(client)
        .then((response) => response.data)
        .catch((error) => {
          console.error(`${error.code}: ${ERROR_MESSAGES.FETCH_CLIENT_FAILED_ERROR}`);
        });
    }),
  );
};

export const codegenFromConfig = () => {
  const codegenConfigPath = path.resolve("ts-codegen.config.json");

  const getCodegenConfig = () =>
    fs.existsSync(codegenConfigPath)
      ? require(codegenConfigPath)
      : {
          output: ".output",
          actionCreatorImport: "",
          clients: [],
        };
  const { output, actionCreatorImport, timeout, data, clients } = getCodegenConfig();

  const writeSpecToFile = (spec: IOpenAPI | Spec) => {
    const fileStr = `${actionCreatorImport} ${codegen(spec)}`;
    write(output, getFilename(spec.basePath), fileStr);
  };

  (data || []).map((file: string) => {
    const specStr = fs.readFileSync(file, "utf8");
    const spec = testJSON(specStr);

    writeSpecToFile(spec);
  });

  if (clients) {
    fetchSwaggerJSON(clients, timeout).then((results: any[]) => {
      results.forEach((spec: IOpenAPI | Spec) => {
        writeSpecToFile(spec);
      });
    });
  }
};

codegenFromConfig();
