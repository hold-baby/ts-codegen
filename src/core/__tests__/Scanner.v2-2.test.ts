import swaggerV2PetStore from "examples/petstore.json";
import { Scanner } from "src/core/Scanner";

describe("swagger v2", () => {
  it("should handle special spec correctly", () => {
    expect(new Scanner(swaggerV2PetStore as any).scan()).toMatchSnapshot();
  });
});
