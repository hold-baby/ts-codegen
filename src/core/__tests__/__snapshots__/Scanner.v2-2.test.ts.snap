// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`swagger v2 should handle special spec correctly 1`] = `
"export const addPet = createRequestAction<{
  requestBody: {
    body?: IPet;
  };
}>(\\"addPet\\", ({ requestBody }) => ({
  url: \`/v2/pet\`,
  method: \\"post\\",
  data: requestBody,
  headers: { \\"Content-Type\\": \\"application/json\\" },
}));

export const createUser = createRequestAction<{
  requestBody: {
    body?: IUser;
  };
}>(\\"createUser\\", ({ requestBody }) => ({
  url: \`/v2/user\`,
  method: \\"post\\",
  data: requestBody,
  headers: { \\"Content-Type\\": \\"application/json\\" },
}));

export const createUsersWithArrayInput = createRequestAction<{
  requestBody: {
    body?: IUser[];
  };
}>(\\"createUsersWithArrayInput\\", ({ requestBody }) => ({
  url: \`/v2/user/createWithArray\`,
  method: \\"post\\",
  data: requestBody,
  headers: { \\"Content-Type\\": \\"application/json\\" },
}));

export const createUsersWithListInput = createRequestAction<{
  requestBody: {
    body?: IUser[];
  };
}>(\\"createUsersWithListInput\\", ({ requestBody }) => ({
  url: \`/v2/user/createWithList\`,
  method: \\"post\\",
  data: requestBody,
  headers: { \\"Content-Type\\": \\"application/json\\" },
}));

export const deleteOrder = createRequestAction<{
  orderId: string;
}>(\\"deleteOrder\\", ({ orderId }) => ({ url: \`/v2/store/order/\${orderId}\`, method: \\"delete\\" }));

export const deletePet = createRequestAction<{
  petId: number;
}>(\\"deletePet\\", ({ petId }) => ({ url: \`/v2/pet/\${petId}\`, method: \\"delete\\" }));

export const deleteUser = createRequestAction<{
  username: string;
}>(\\"deleteUser\\", ({ username }) => ({ url: \`/v2/user/\${username}\`, method: \\"delete\\" }));

export const findPetsByStatus = createRequestAction<
  {
    status?: string[];
  },
  IPet[]
>(\\"findPetsByStatus\\", ({ status }) => ({
  url: \`/v2/pet/findByStatus\`,
  method: \\"get\\",
  params: {
    status,
  },
}));

export const findPetsByTags = createRequestAction<
  {
    tags?: string[];
  },
  IPet[]
>(\\"findPetsByTags\\", ({ tags }) => ({
  url: \`/v2/pet/findByTags\`,
  method: \\"get\\",
  params: {
    tags,
  },
}));

export const getInventory = createRequestAction<{ [key: string]: any }>(\\"getInventory\\", () => ({
  url: \`/v2/store/inventory\`,
  method: \\"get\\",
}));

export const getOrderById = createRequestAction<
  {
    orderId: string;
  },
  IOrder
>(\\"getOrderById\\", ({ orderId }) => ({ url: \`/v2/store/order/\${orderId}\`, method: \\"get\\" }));

export const getPetById = createRequestAction<
  {
    petId: number;
  },
  IPet
>(\\"getPetById\\", ({ petId }) => ({ url: \`/v2/pet/\${petId}\`, method: \\"get\\" }));

export const getUserByName = createRequestAction<
  {
    username: string;
  },
  IUser
>(\\"getUserByName\\", ({ username }) => ({ url: \`/v2/user/\${username}\`, method: \\"get\\" }));

export const loginUser = createRequestAction<
  {
    password?: string;
    username?: string;
  },
  string
>(\\"loginUser\\", ({ username, password }) => ({
  url: \`/v2/user/login\`,
  method: \\"get\\",
  params: {
    username,
    password,
  },
}));

export const logoutUser = createRequestAction(\\"logoutUser\\", () => ({ url: \`/v2/user/logout\`, method: \\"get\\" }));

export const placeOrder = createRequestAction<
  {
    requestBody: {
      body?: IOrder;
    };
  },
  IOrder
>(\\"placeOrder\\", ({ requestBody }) => ({
  url: \`/v2/store/order\`,
  method: \\"post\\",
  data: requestBody,
  headers: { \\"Content-Type\\": \\"application/json\\" },
}));

export const updatePet = createRequestAction<{
  requestBody: {
    body?: IPet;
  };
}>(\\"updatePet\\", ({ requestBody }) => ({
  url: \`/v2/pet\`,
  method: \\"put\\",
  data: requestBody,
  headers: { \\"Content-Type\\": \\"application/json\\" },
}));

export const updatePetWithForm = createRequestAction<{
  petId: string;
  requestBody: {
    name?: string;
    status?: string;
  };
}>(\\"updatePetWithForm\\", ({ petId, requestBody }) => ({
  url: \`/v2/pet/\${petId}\`,
  method: \\"post\\",
  data: requestBody,
  headers: { \\"Content-Type\\": \\"multipart/form-data\\" },
}));

export const updateUser = createRequestAction<{
  requestBody: {
    body?: IUser;
  };
  username: string;
}>(\\"updateUser\\", ({ username, requestBody }) => ({
  url: \`/v2/user/\${username}\`,
  method: \\"put\\",
  data: requestBody,
  headers: { \\"Content-Type\\": \\"application/json\\" },
}));

export const uploadFile = createRequestAction<{
  petId: number;
  requestBody: {
    additionalMetadata?: string;
    file?: File;
  };
}>(\\"uploadFile\\", ({ petId, requestBody }) => ({
  url: \`/v2/pet/\${petId}/uploadImage\`,
  method: \\"post\\",
  data: requestBody,
  headers: { \\"Content-Type\\": \\"multipart/form-data\\" },
}));

export interface ICategory {
  id?: number;
  name?: string;
}

export interface IOrder {
  complete?: boolean;
  id?: number;
  petId?: number;
  quantity?: number;
  shipDate?: string;
  status?: keyof typeof OrderStatus;
}

export enum OrderStatus {
  \\"placed\\" = \\"placed\\",
  \\"approved\\" = \\"approved\\",
  \\"delivered\\" = \\"delivered\\",
}

export interface IPet {
  category?: ICategory;
  id?: number;
  name: string;
  photoUrls: string[];
  status?: keyof typeof PetStatus;
  tags?: ITag[];
}

export enum PetStatus {
  \\"available\\" = \\"available\\",
  \\"pending\\" = \\"pending\\",
  \\"sold\\" = \\"sold\\",
}

export interface ITag {
  id?: number;
  name?: string;
}

export interface IUser {
  email?: string;
  firstName?: string;
  id?: number;
  lastName?: string;
  password?: string;
  phone?: string;
  userStatus?: number;
  username?: string;
}
"
`;
